### PR TITLE
sc-57182/fix-date-masking

### PR DESCRIPTION
**Pull request by johnrb2**
## Description of the change

> fix logic for masking dates

## Type of change
- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Configuration change
- [ ] Technical Debt
- [ ] Documentation

## Related tickets

> https://app.shortcut.com/active-prospect/story/57182/fix-date-masking

### Development and Testing

- [x]  Lint rules pass locally.
- [x]  The code changed/added as part of this pull request has been covered with tests, or this PR does not alter production code.
- [x]  All tests related to the changed code pass in development, or tests are not applicable.

### Code Review

- [x]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached.
- [x]  At least two engineers have been added as "Reviewers" on the pull request.
- [ ]  Changes have been reviewed by at least two other engineers who did not write the code.
- [x]  This branch has been rebased off master to be current.

### Tracking 
- [x]  Issue from Shortcut/Jira has a link to this pull request.
- [x]  This PR has a link to the issue in Shortcut.

### QA
- [ ]  This branch has been deployed to staging and tested.

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -40,15 +40,15 @@ const mask = function (obj, doMask = false) {
     obj = obj.map(i => mask(i, doMask));
   } else if (typeof obj === 'object') {
     // a Date object with an age property is a dob type field
-    if (obj.masked === false && obj instanceof Date && obj.age) {
+    if ((doMask || obj.masked === false) && obj instanceof Date && obj.age) {
```
Comment (samullen) [inline lib/index.js:43]: I think this and the conditional on line 51 need to be extracted into functions so people don't have to think about what's going on. Maybe this function's name would be `shouldMaskAge`

File: lib/index.js [hunk 2]
```diff
@@ -40,15 +40,15 @@ const mask = function (obj, doMask = false) {
...
       const str = new String(`**/**/${obj.getFullYear()}`);
       for (const key in obj) {
         if (!_.isFunction(obj[key])) str[key] = obj[key];
       }
       str.raw = `**/**/${obj.getFullYear()}`;
       str.masked = true;
       obj = str;
-    } else if ((obj.masked === false) && (obj instanceof String || obj instanceof Boolean || obj instanceof Number || obj instanceof Date)) {
+    } else if ((doMask || obj.masked === false) && (obj instanceof String || obj instanceof Boolean || obj instanceof Number || obj instanceof Date)) {
```
Comment (samullen) [inline lib/index.js:51]: Maybe extract this to a `shouldMaskStdType`? Probably a better name. Ask ChatGPT. :)

**Review by samullen — COMMENTED** [review]
Couple requests to extract compound conditionals, but looks good otherwise.

File: test/mask.test.js [hunk 3]
```diff
@@ -41,30 +64,33 @@ describe('Mask utility', function () {
...
 
   it('should mask array', function () {
     const masked = mask({
       masked: false,
-      array: [1, '23', true]
+      array: [1, '23', true, date.parse('06/02/2014')]
     });
 
-    assert.deepEqual(masked, {
+    assert.deepEqual(JSON.parse(JSON.stringify(masked)), {
```
Comment (cgrayson) [inline test/mask.test.js:91]: Curious why the `parse(stringify())` is needed here (& on L72) now.

File: lib/index.js [hunk 4]
```diff
@@ -58,7 +58,7 @@ const mask = function (obj, doMask = false) {
           if (!key.match(digit)) { str[key] = mask(value, true); }
         }
       }
-      str.masked = true;
+      if (obj.masked != null) { str.masked = true; }
```
Comment (cgrayson) [inline lib/index.js:61]: I think I see why one of the new tests needed this change, but I wonder if there could be unintended side-effects from it.

File: lib/index.js (hunk 1 above)
Comment (johnrb2) [reply lib/index.js:43]: I think I will only partially extract the logic, since this will be removed in the other DOB story

**Pull request closed without merging**